### PR TITLE
Store win url in cache

### PR DESCRIPTION
**Pull request by AndriyPavlyuk**
Updating event reporting for more reliable work of Prebid Server.

### CODE REVIEW

File: src/main/java/org/prebid/server/cache/CacheService.java [hunk 1]
```diff
@@ -381,10 +382,14 @@ private static PutObject createPutObjectVideoOnly(Bid bid) {
     /**
      * Makes JSON type {@link PutObject} from {@link com.iab.openrtb.response.Bid}. Used for OpenRTB auction request.
      */
-    private PutObject createJsonPutObjectOpenrtb(CacheBid cacheBid) {
+    private PutObject createJsonPutObjectOpenrtb(CacheBid cacheBid, String accountId) {
+        final ObjectNode value = mapper.mapper().valueToTree(cacheBid.getBid());
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/cache/CacheService.java]: change to bidObjectNode

File: src/main/java/org/prebid/server/cache/CacheService.java [hunk 2]
```diff
@@ -381,10 +382,14 @@ private static PutObject createPutObjectVideoOnly(Bid bid) {
     /**
      * Makes JSON type {@link PutObject} from {@link com.iab.openrtb.response.Bid}. Used for OpenRTB auction request.
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/cache/CacheService.java]: And add wurl

File: src/test/java/org/prebid/server/cache/CacheServiceTest.java [hunk 3]
```diff
@@ -458,11 +468,13 @@ public void cacheBidsOpenrtbShouldTolerateCacheEntriesNumberDoesNotMatchBidsNumb
     public void cacheBidsOpenrtbShouldReturnExpectedDebugInfo() throws JsonProcessingException {
         // given
         final com.iab.openrtb.response.Bid bid = givenBidOpenrtb(builder -> builder.id("bidId1").impid("impId1"));
-
+        given(eventsService.winUrlTargeting(anyString())).willReturn("http://win-url");
+        final ObjectNode value = mapper.valueToTree(bid);
+        value.put("wurl","http://win-url" );
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/cache/CacheServiceTest.java]:
``` final ObjectNode value = mapper.valueToTree(bid);
    value.put("wurl","http://win-url" );```
Are they really required?

File: src/test/java/org/prebid/server/cache/CacheServiceTest.java [hunk 4]
```diff
@@ -438,11 +446,13 @@ public void cacheBidsOpenrtbShouldTolerateCacheEntriesNumberDoesNotMatchBidsNumb
         givenHttpClientReturnsResponse(200, "{}");
 
         final com.iab.openrtb.response.Bid bid = givenBidOpenrtb(builder -> builder.id("bidId1").impid("impId1"));
-
+        given(eventsService.winUrlTargeting(anyString())).willReturn("http://win-url");
+        final ObjectNode value = mapper.valueToTree(bid);
+        value.put("wurl","http://win-url" );
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/cache/CacheServiceTest.java]:
``` final ObjectNode value = mapper.valueToTree(bid);
        value.put("wurl","http://win-url" );```
Are they really required?

File: src/test/java/org/prebid/server/cache/CacheServiceTest.java [hunk 5]
```diff
@@ -494,18 +506,23 @@ public void cacheBidsOpenrtbShouldPerformHttpRequestWithExpectedBody() throws IO
...
         cacheService.cacheBidsOpenrtb(
                 asList(bid1, bid2), asList(imp1, imp2),
-                CacheContext.builder().shouldCacheBids(true).shouldCacheVideoBids(true).build(), account, timeout);
+                CacheContext.builder().shouldCacheBids(true).shouldCacheVideoBids(true).build(), Account.builder().id("accountId").build(), timeout);
 
         // then
+        final ObjectNode value1 = mapper.valueToTree(bid1);
+        value1.put("wurl", "http://win-url");
+        final ObjectNode value2 = mapper.valueToTree(bid2);
+        value2.put("wurl", "http://win-url");
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/cache/CacheServiceTest.java]: replace naming of value1 and value2

File: src/test/java/org/prebid/server/cache/CacheServiceTest.java [hunk 6]
```diff
@@ -494,18 +506,23 @@ public void cacheBidsOpenrtbShouldPerformHttpRequestWithExpectedBody() throws IO
         final com.iab.openrtb.response.Bid bid2 = givenBidOpenrtb(builder -> builder.impid("impId2").adm("adm2"));
         final Imp imp1 = givenImp(identity());
         final Imp imp2 = givenImp(builder -> builder.id("impId2").video(Video.builder().build()));
+        when(eventsService.winUrlTargeting(anyString())).thenReturn("http://win-url");
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/cache/CacheServiceTest.java]: Replace with `given` and `willReturn`

File: src/test/java/org/prebid/server/cache/CacheServiceTest.java [hunk 7]
```diff
@@ -721,18 +738,23 @@ public void cacheBidsOpenrtbShouldWrapEmptyAdMFieldUsingNurlFieldValue() throws
...
         cacheService.cacheBidsOpenrtb(
                 asList(bid1, bid2), singletonList(imp1),
-                CacheContext.builder().shouldCacheBids(true).shouldCacheVideoBids(true).build(), account, timeout);
+                CacheContext.builder().shouldCacheBids(true).shouldCacheVideoBids(true).build(), Account.builder().id("accountId").build(), timeout);
 
         // then
+        final ObjectNode value1 = mapper.valueToTree(bid1);
+        value1.put("wurl", "http://win-url");
+        final ObjectNode value2 = mapper.valueToTree(bid2);
+        value2.put("wurl", "http://win-url");
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/cache/CacheServiceTest.java]: naming and inline

File: src/test/java/org/prebid/server/cache/CacheServiceTest.java [hunk 8]
```diff
@@ -748,16 +770,19 @@ public void cacheBidsOpenrtbShouldNotModifyVastXmlWhenBidIdIsNotInToModifyList()
                 .adm("adm"));
         final Imp imp1 = givenImp(builder -> builder.id("impId1").video(Video.builder().build()));
 
+        given(eventsService.winUrlTargeting(anyString())).willReturn("http://win-url");
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/cache/CacheServiceTest.java]: /n

File: src/test/java/org/prebid/server/cache/CacheServiceTest.java [hunk 9]
```diff
@@ -767,17 +792,18 @@ public void cacheBidsOpenrtbShouldNotAddTrackingImpToBidAdmWhenXmlDoesNotContain
         final com.iab.openrtb.response.Bid bid = givenBidOpenrtb(builder -> builder.id("bid1").impid("impId1")
                 .adm("no impression tag"));
         final Imp imp1 = givenImp(builder -> builder.id("impId1").video(Video.builder().build()));
-
+        given(eventsService.winUrlTargeting(anyString())).willReturn("http://win-url");
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/cache/CacheServiceTest.java]: /n

File: src/test/java/org/prebid/server/cache/CacheServiceTest.java [hunk 10]
```diff
@@ -767,17 +792,18 @@ public void cacheBidsOpenrtbShouldNotAddTrackingImpToBidAdmWhenXmlDoesNotContain
...
         // when
         cacheService.cacheBidsOpenrtb(singletonList(bid), singletonList(imp1), CacheContext.builder()
                 .shouldCacheBids(true).shouldCacheVideoBids(true).videoBidIdsToModify(singletonList("bid1"))
-                .build(), account, timeout);
-
+                .build(), Account.builder().id("accountId").build(), timeout);
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/cache/CacheServiceTest.java]: Extract variables to clean code

File: src/test/resources/org/prebid/server/it/openrtb2/rubicon_appnexus/test-cache-rubicon-appnexus-request.json [hunk 11]
```diff
@@ -17,7 +18,8 @@
         "adm": "adm2",
         "crid": "crid2",
         "w": 300,
-        "h": 600
+        "h": 600,
+        "wurl": "http://localhost:8000/event?t=win&b=BIDID&a=5001&f=i"
```
Comment (DGarbar) [inline src/test/resources/org/prebid/server/it/openrtb2/rubicon_appnexus/test-cache-rubicon-appnexus-request.json]: Please clarify with Ruslan, if this tests will not fail on another environment bc of `http://localhost:8000/`

File: src/main/java/org/prebid/server/cache/CacheService.java [hunk 12]
```diff
@@ -379,12 +380,18 @@ private static PutObject createPutObjectVideoOnly(Bid bid) {
     }
 
     /**
-     * Makes JSON type {@link PutObject} from {@link com.iab.openrtb.response.Bid}. Used for OpenRTB auction request.
+     * Makes JSON type {@link PutObject} from {@link com.iab.openrtb.response.Bid}. Used for OpenRTB auction request and
+     * add winn url.
      */
-    private PutObject createJsonPutObjectOpenrtb(CacheBid cacheBid) {
+    private PutObject createJsonPutObjectOpenrtb(CacheBid cacheBid, String accountId) {
+        final String bidId = cacheBid.getBid().getId();
```
Comment (rpanchyk) [inline src/main/java/org/prebid/server/cache/CacheService.java]: Pls extract `cacheBid.getBid()` to separate variable.
Then `bidId` declaration can be omitted.

File: src/test/java/org/prebid/server/cache/CacheServiceTest.java [hunk 13]
```diff
@@ -396,8 +404,9 @@ public void cacheBidsOpenrtbShouldTolerateResponseCodeIsNot200() throws JsonProc
         givenHttpClientReturnsResponse(503, "response");
 
         final com.iab.openrtb.response.Bid bid = givenBidOpenrtb(builder -> builder.id("bidId1").impid("impId1"));
-
+        given(eventsService.winUrlCaching(anyString(), anyString())).willReturn("http://win-url");
         // when
```
Comment (rpanchyk) [inline src/test/java/org/prebid/server/cache/CacheServiceTest.java]: Minor. Pls add new line.

File: src/test/resources/org/prebid/server/it/openrtb2/rubicon_appnexus/test-cache-rubicon-appnexus-request.json (hunk 11 above)
Comment (rpanchyk) [reply src/test/resources/org/prebid/server/it/openrtb2/rubicon_appnexus/test-cache-rubicon-appnexus-request.json]: That's fine. Will refactor in future.

File: src/main/java/org/prebid/server/events/EventsService.java [hunk 14]
```diff
@@ -31,6 +31,13 @@ public String winUrlTargeting(String accountId) {
         return eventUrl(EventRequest.Type.win, BIDID_PLACEHOLDER, accountId, EventRequest.Format.image);
     }
 
+    /**
+     * Returns value for caching winurl keyword in request.
```
Comment (rpanchyk) [inline src/main/java/org/prebid/server/events/EventsService.java]: Pls change to `Returns value for win URL tracking.`

File: src/main/java/org/prebid/server/events/EventsService.java [hunk 15]
```diff
@@ -31,6 +31,13 @@ public String winUrlTargeting(String accountId) {
...
+     */
+    public String winUrlCaching(String bidId, String accountId) {
```
Comment (rpanchyk) [inline src/main/java/org/prebid/server/events/EventsService.java]: Rename to `winUrl(...)`.

File: src/test/java/org/prebid/server/cache/CacheServiceTest.java [hunk 16]
```diff
@@ -370,12 +370,28 @@ public void cacheBidsOpenrtbShouldTolerateGlobalTimeoutAlreadyExpired() {
         assertThat(result.getHttpCall()).isNull();
     }
 
+    @Test
+    public void cacheBidsOpenrtbShouldStoreWinUrl() throws JsonProcessingException {
+        // given
+        final com.iab.openrtb.response.Bid bid = givenBidOpenrtb(builder -> builder.id("bidId1").impid("impId1"));
+
+        // when
+        account = Account.builder().id("accountId").build();
```
Comment (rpanchyk) [inline src/test/java/org/prebid/server/cache/CacheServiceTest.java]: This declaration should be a part of `given` section.
Pls check all remaining tests.

File: src/test/java/org/prebid/server/cache/CacheServiceTest.java [hunk 17]
```diff
@@ -721,18 +746,24 @@ public void cacheBidsOpenrtbShouldWrapEmptyAdMFieldUsingNurlFieldValue() throws
         final com.iab.openrtb.response.Bid bid1 = givenBidOpenrtb(builder -> builder.impid("impId1").adm("adm1"));
         final com.iab.openrtb.response.Bid bid2 = givenBidOpenrtb(builder -> builder.impid("impId1").nurl("adm2"));
         final Imp imp1 = givenImp(builder -> builder.id("impId1").video(Video.builder().build()));
+        given(eventsService.winUrlCaching(any(), any())).willReturn("http://win-url");
+        given(eventsService.winUrlCaching(any(), any())).willReturn("http://win-url");
```
Comment (rpanchyk) [inline src/test/java/org/prebid/server/cache/CacheServiceTest.java]: No reason for duplicate.

**Pull request merged**